test/test-execute/exec-temporaryfilesystem-ro.service

```diff
@@ -0,0 +1,33 @@
+[Unit]
+Description=Test for TemporaryFileSystem with options
+
+[Service]
+Type=oneshot
+
+# Check directories exist
+ExecStart=/bin/sh -c 'test -d /var/test-exec-temporaryfilesystem-rw && test -d /var/test-exec-temporaryfilesystem-ro'
+
+# Check TemporaryFileSystem= are empty
+ExecStart=/bin/sh -c 'for i in $$(ls -A /var); do test $$i = test-exec-temporaryfilesystem-rw -o $$i = test-exec-temporaryfilesystem-ro || false; done'
+
+# Cannot create a file in /var
+ExecStart=/bin/sh -c '! touch /var/hoge'
+
+# Create a file in /var/test-exec-temporaryfilesystem-rw
+ExecStart=/bin/sh -c 'touch /var/test-exec-temporaryfilesystem-rw/thisisasimpletest-temporaryfilesystem'
+
+# Then, the file can be access through /tmp
+ExecStart=/bin/sh -c 'test -f /tmp/thisisasimpletest-temporaryfilesystem'
+
+# Also, through /var/test-exec-temporaryfilesystem-ro
+ExecStart=/bin/sh -c 'test -f /var/test-exec-temporaryfilesystem-ro/thisisasimpletest-temporaryfilesystem'
+
+# The file cannot modify through /var/test-exec-temporaryfilesystem-ro
+ExecStart=/bin/sh -c '! touch /var/test-exec-temporaryfilesystem-ro/thisisasimpletest-temporaryfilesystem'
+
+# Cleanup
+ExecStart=/bin/sh -c 'rm /tmp/thisisasimpletest-temporaryfilesystem'
+
+TemporaryFileSystem=/var:ro
+BindPaths=/tmp:/var/test-exec-temporaryfilesystem-rw
```

Comment:
Sorry, another review dribble.  I noticed some subtle points by looking at this.

## Point 1

I think you pointed earlier that this is able to work correctly because we remount readonly in a second step.  I.e. only after we created /var/test-exec-temporaryfilesystem-rw, do we remount /var as readonly.

It's a pity the "add documents" commit would not let users realize that this works.  And this mode of operation is the ideal one for the original use case.  Maybe including an example of it in the documentation would be enough, at least to encourage users to try it and see that it works :).

## Not so subtle point 2

The original examples were like allowing access to /etc/bind but not allowing access to anything else in /etc.  This is quite subtle to do (I think there's similar comments elsewhere in the code), because if systemd first overmounts /etc, then it cannot easily access /etc/bind in order to bind-mount it.  If this works, we really really want it to be tested.  If this doesn't work, I think we don't want to close those bugs yet?

I had an idea earlier, to effectively enable the same code used for `RootDirectory=` even when the root directory is not different.  (Though it should be skipped when we're not doing _any_ namespace games).  `mount --rbind / /run/systemd/unit-root` is less expensive than you might think.  But it gets more complex, because currently we just use `mount --move /run/systemd/unit-root /`.  That means we leave _everything_ overmounted, it should work fine but it's ugly.  Ideally we would use `cd /run/systemd/unit-root && pivot_root . run/systemd/old-root`, and then `umount -l /run/systemd/old-root`.  (I originally tried umount -R /run/systemd/old-root` but [https://unix.stackexchange.com/questions/418100/umount-r-on-bind-mounts-takes-a-non-neglible-amount-of-time-why](found it caused an unexpected delay)).  I think my original motivation for this scheme was different and maybe sounds silly: if setup_namespace() fails, we currently ignore it and continue with the half-setup namespace.  My scheme would allow setup_namespace() to clean up with a simple `umount -l /run/systemd/unit-root`.

I had a WIP branch for this which consists of one code comment and two prep commits :).  https://github.com/sourcejedi/systemd/tree/rootdirectory

## Point 3

There's a mismatch between what the current remount-readonly code does, and what we would have written for this case specifically.  The current code is _bind_mounting the tmpfs read-only.  This is legal, it doesn't create an extra mount entry, and it we probably don't care about the difference.  As well as affecting the behaviour when it is bind mounted, the difference will be visible in /proc/self/mountinfo

    449 442 0:48 / /tmp ro,relatime - tmpfs tmpfs rw,seclabel

as opposed to

    448 443 0:48 / /mnt ro,relatime - tmpfs tmpfs ro,seclabel

Reply:
My scheme in point 2 has the caveat that the `pivot_root()` syscall implementation is a bit funky.  The manpage for it points out that it involves scanning kernel threads to check if any of them need changing directory.  So I'm not 100% confident whether or not it's a good idea to call `pivot_root()` for anything other than switching from initramfs to the real system.

Reply:
About point 3. With exec-temporaryfilesystem-ro.service, I get the following.
```
# /proc/self/mountinfo
438 400 0:58 / /var ro,nodev shared:226 - tmpfs tmpfs rw
# /proc/self/mounts
tmpfs /var tmpfs ro,nodev 0 0
```
It may be misleading. But, at least, this seems not in the scope of this PR, I think.

Reply:
About point 2. Sorry, I have no idea about this... But your point seems about whole design of `setup_namespace()` rather than the new setting `TemporaryFileSystem=`. Right? If so, this is out of the scope this PR provides, I think.

Reply:
> It may be misleading. But, at least, this seems not in the scope of this PR, I think, as this 'problem' affects not only TemporaryFileSystem= but also ReadOnlyPaths= or BindReadOnlyPaths=. Moreover, of course, I do not say that we do not need to fix this, but, unfortunately, I do not know how to fix this.

Thank you for looking.

1) I don't think we need to fix this.  I thought you might want to know though.  I'd still say it looks like the code _could_ have been written deliberately.

2) "Misleading" sounds a little strong.  That said, it doesn't 100% match the documentation.  I.e. when we set `ro` as a mount option, then "set as a mount option" is meaning something slightly different than for all the other mount options.

3) ReadOnlyPaths= and BindReadOnlyPaths= are both bind mounts, so it is expected that they work like `mount -o bind,ro olddir newdir`.  They cannot work any other way.

However TemporaryFileSystem= is a tmpfs mount.  So we have the opportunity of setting options on the underlying filesystem ("superblock"), instead of on the individual mount point.

In an ideal world, we would do the exact right thing, even on the obscure detail.  It would mean that if you play around, discover the obscure detail, and become confused - then your confusion will be because systemd did the right thing, not the wrong thing :).  And you will be happy to learn and have no annoyance at systemd :).

Reply:
> About point 2. Sorry, I have no idea about this... But your point seems about whole design of setup_namespace() rather than the new setting TemporaryFileSystem=. Right? If so, this is out of the scope this PR provides, I think.

You need to have an idea about this, sorry :).  Please can you test the case by @topimiettinen 

    TemporaryFileSystem=/usr
    ReadOnlyPaths=/usr/lib

Can the service can see the full contents of /usr/lib?  If so, please make sure there is a test case that covers it.

If it doesn't work, you need to change the text of this PR to make sure that it does not close issues #7895, #7153, and #2780.

As you say, this could be addressed by a different PR.  (TemporaryFileSystem= could still be used for some purpose, but in a different way, matching how systemd uses it for the service private directories).  Sorry for making such a confusing comment about setup_namespace() :).  I am very pleased with my idea :) but the best explanation would be if I had working code to show you.

Reply:
I am very sympathetic with the amount of work you have here, and not wanting further delays from out-of-scope work :).

Reply:
For the httpd use case, should this work? 
```
ProtectHome=tmpfs
ReadOnlyPaths=/home/httpd
```